### PR TITLE
OS-7746 EFI text console isn't working when booted via iPXE

### PR DESCRIPTION
**Pull request by tsoome**
Add multiboot_tag_framebuffer setup for multiboot2. This information is needed for UEFI local console implementation in illumos kernel.

### CODE REVIEW

**Review by bahamat — DISMISSED** [review]
This looks good to me. Maybe should have Dan look at it also. We have 3 weeks until the next Triton release, so we've got time.

**Review by danmcd — COMMENTED** [review]
My initial stab at this put things in other files.  I'm liking you managing to keep it encapsulated in multiboot2.c at first glance.  ESPECIALLY given this will never make it upstream.

File: src/arch/x86/image/multiboot2.c [hunk 1]
```diff
@@ -461,6 +463,173 @@ static int multiboot2_add_bootloader ( struct mb2 *mb2 ) {
 	return 0;
 }
 
+static EFI_STATUS efifb_colour_map_mask ( uint32_t mask, uint8_t *scale,
```
Comment (danmcd) [inline src/arch/x86/image/multiboot2.c]: Somewhere I'm fighting against imaginary Josh Clulow about the spelling of "color" or "colour". Keep it.

File: src/arch/x86/image/multiboot2.c [hunk 2]
```diff
@@ -461,6 +463,173 @@ static int multiboot2_add_bootloader ( struct mb2 *mb2 ) {
...
+	/* Check that original mask was contiguous */
+	check = ( ( 0xff >> ( 8 - *scale ) ) << *lsb );
+	if ( check != mask )
+		return EFI_UNSUPPORTED;
+
+	return EFI_SUCCESS;
+}
+
+#if !defined(EFI_CONSOLE_OUT_DEVICE_GUID)
+#define	EFI_CONSOLE_OUT_DEVICE_GUID	\
```
Comment (danmcd) [inline src/arch/x86/image/multiboot2.c:486]: Could you cite a source on this?  E.g.

https://uefi.org/specs/UEFI/2.9_A/12_Protocols_Console_Support.html

File: src/arch/x86/image/multiboot2.c [hunk 3]
```diff
@@ -461,6 +463,173 @@ static int multiboot2_add_bootloader ( struct mb2 *mb2 ) {
...
+
+	/* Get all GOP handles */
+	status = bs->LocateHandleBuffer ( ByProtocol,
+		&efi_graphics_output_protocol_guid, NULL, &count, &handles );
+	if (status != EFI_SUCCESS)
+		return -ENOTSUP;
+
+	/*
+	 * Search for ConOut protocol, if not found, use first handle.
+	 * See illumos issue "13453 loader.efi: handle multiple gop instances"
```
Comment (danmcd) [inline src/arch/x86/image/multiboot2.c:507]: This inspired my prior citing-sources comment.

**Review by danmcd — COMMENTED** [review]
Mere nits is all.

File: src/arch/x86/image/multiboot2.c (hunk 1 above)
Comment (tsoome) [reply src/arch/x86/image/multiboot2.c]: > Somewhere I'm fighting against imaginary Josh Clulow about the spelling of "color" or "colour". Keep it.

its colour all over the source there:) actually, I think I did create confusion there -- in efi_fbcon.c there is already local function with this name. I better rename it.

File: src/arch/x86/image/multiboot2.c (hunk 2 above)
Comment (tsoome) [reply src/arch/x86/image/multiboot2.c:486]: It is actually complicated -- this specific GUID is not universally available, hence we need to keep alternate option (and not error out when not available). But it was found out usable at least on some hardware.

**Review by danmcd — APPROVED** [review]
Thanks for the background.

I asked a question on the ticket about testing results.  Once answered I can IA as well.

**Comment by danmcd** [comment]
Will grant IA with a second reviewer here.

File: src/arch/x86/image/multiboot2.c (hunk 1 above)
Comment (bahamat) [reply src/arch/x86/image/multiboot2.c]: I'm in favor of sticking with consistent spelling.

**Review by bahamat — APPROVED** [review]

**Pull request merged**